proxy/core/src/main/scala/io/cloudstate/proxy/eventsourced/InMemJournal.scala

```diff
@@ -0,0 +1,220 @@
+      case unsupported =>
+        throw new IllegalArgumentException(s"$unsupported is an unsupported offset type for the in memory read journal")
+    }
+
+    Source
+      .fromFutureSource(
+        (inmemJournal ? EventsByTag(tag, fromOffset)).mapTo[Source[EventEnvelope, NotUsed]]
+      )
+      .mapMaterializedValue(_ => NotUsed)
+  }
+}
+
+/**
+ * INTERNAL API.
+ */
+trait InmemMessages {
+  private var messages = Map.empty[String, Vector[Message]]
+  private var allMessages = Vector.empty[Message]
+  private var offset: Long = 0
+
+  private val (broadcast, source) = Source
+    .actorRef[Message](8, OverflowStrategy.fail)
+    .toMat(BroadcastHub.sink)(Keep.both)
+    .run()
+
+  protected implicit def mat: Materializer
+
+  def add(p: PersistentRepr): Unit = {
+    val message = toMessage(p)
+
+    messages = messages + (messages.get(p.persistenceId) match {
+        case Some(ms) => p.persistenceId -> (ms :+ message)
+        case None => p.persistenceId -> Vector(message)
+      })
+
+    allMessages :+= message
+    broadcast ! message
+  }
+
+  def update(pid: String, snr: Long)(f: PersistentRepr => PersistentRepr): Unit = {
+    messages = messages.get(pid) match {
+      case Some(ms) =>
+        messages + (pid -> ms.map(sp => if (sp.repr.sequenceNr == snr) sp.copy(repr = f(sp.repr)) else sp))
+      case None => messages
+    }
+    allMessages = allMessages.map { message =>
+      if (message.repr.persistenceId == pid && message.repr.sequenceNr == snr) {
+        // todo can tags change?
+        message.copy(repr = f(message.repr))
+      } else {
+        message
+      }
+    }
+  }
+
+  def delete(pid: String, snr: Long): Unit = {
+    messages = messages.get(pid) match {
+      case Some(ms) =>
+        messages + (pid -> ms.filterNot(_.repr.sequenceNr == snr))
+      case None => messages
+    }
+    allMessages = allMessages.filterNot { message =>
+      message.repr.persistenceId == pid && message.repr.sequenceNr == snr
+    }
+  }
+
+  def read(pid: String, fromSnr: Long, toSnr: Long, max: Long): immutable.Seq[PersistentRepr] =
+    messages.get(pid) match {
+      case Some(ms) =>
+        ms.view
+          .dropWhile(_.repr.sequenceNr < fromSnr)
+          .takeWhile(_.repr.sequenceNr <= toSnr)
+          .take(safeLongToInt(max))
+          .map(_.repr)
+          .toSeq
+      case None => Nil
+    }
+
+  def highestSequenceNr(pid: String): Long = {
+    val snro = for {
+      ms <- messages.get(pid)
+      m <- ms.lastOption
+    } yield m.repr.sequenceNr
+    snro.getOrElse(0L)
+  }
+
+  def eventsByTagQuery(tag: String, fromOffset: Long): Source[EventEnvelope, NotUsed] =
+    // Technically, there's a race condition that could result in messages being dropped here, if a new message is
+    // persisted after this returns, but before the subscriber to the query materializes this stream, that message
+    // will be dropped. But this is only the in memory journal which makes no sense to use in production.
```

Comment:
Fair enough that it's likely not needed. If we did want to close this gap, I guess it could retrieve the old messages more as a lazy source once materialized, or something similar.